drivers/serial/uart_bcm2711.c

```diff
@@ -0,0 +1,352 @@
+
+	/* Setup 8bit data width and baudrate */
+	sys_write32(BCM2711_MU_LCR_8BIT, lcr);
+	if (!skip_baudrate_config) {
+		divider = (input_clock / (baudrate * 8));
+		sys_write32(divider - 1, baud);
+	}
+
+	/* Enable RX & TX port */
+	val = BCM2711_MU_CNTL_RX_ENABLE | BCM2711_MU_CNTL_TX_ENABLE;
+	sys_write32(val, cntl);
+}
+
+/**
+ * @brief Initialize UART channel
+ *
+ * This routine is called to reset the chip in a quiescent state.
+ * It is assumed that this function is called only once per UART.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0
+ */
+static int uart_bcm2711_init(const struct device *dev)
+{
+	const struct bcm2711_uart_config *uart_cfg = (dev)->config;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+	bcm2711_mu_lowlevel_init(DEVICE_MMIO_GET(dev), 1, uart_cfg->baud_rate, uart_cfg->clocks);
+	return 0;
+}
+
+static void uart_bcm2711_poll_out(const struct device *dev, unsigned char c)
+{
+	bcm2711_mu_lowlevel_putc(DEVICE_MMIO_GET(dev), c);
+}
+
+static int uart_bcm2711_poll_in(const struct device *dev, unsigned char *c)
+{
+	while (!bcm2711_mu_lowlevel_can_getc(DEVICE_MMIO_GET(dev)))
+		;
+
+	return (char)(sys_read32((DEVICE_MMIO_GET(dev) + BCM2711_MU_IO)) & 0xFF);
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int uart_bcm2711_fifo_fill(const struct device *dev,
+				  const uint8_t *tx_data,
+				  int size)
+{
+	unsigned int num_tx = 0U;
+
+	while ((size - num_tx) > 0) {
+		/* Send a character */
+		bcm2711_mu_lowlevel_putc(DEVICE_MMIO_GET(dev), tx_data[num_tx]);
+		num_tx++;
+	}
+
+	return (int)num_tx;
+}
+
+static int uart_bcm2711_fifo_read(const struct device *dev, uint8_t *rx_data,
+				  const int size)
+{
+	unsigned int num_rx = 0U;
+
+	while ((size - num_rx) > 0) {
+		/* Receive a character */
+		rx_data[num_rx++] = bcm2711_mu_lowlevel_getc(DEVICE_MMIO_GET(dev));
+	}
+	return num_rx;
+}
+
+static void uart_bcm2711_irq_tx_enable(const struct device *dev)
+{
+	uint32_t val = 0;
+	mem_addr_t cntl = (DEVICE_MMIO_GET(dev) + BCM2711_MU_CNTL);
+
+	val = BCM2711_MU_CNTL_TX_ENABLE;
+	sys_write32(val, cntl);
+}
+
+static void uart_bcm2711_irq_tx_disable(const struct device *dev)
+{
+	uint32_t val = 0;
+	mem_addr_t cntl = (DEVICE_MMIO_GET(dev) + BCM2711_MU_CNTL);
+
+	val = (~BCM2711_MU_CNTL_TX_ENABLE);
+	sys_write32(val, cntl);
+}
+
+static int uart_bcm2711_irq_tx_ready(const struct device *dev)
+{
+	return bcm2711_mu_lowlevel_can_putc(DEVICE_MMIO_GET(dev));
+}
+
+static void uart_bcm2711_irq_rx_enable(const struct device *dev)
+{
+	uint32_t val = 0;
+	mem_addr_t cntl = (DEVICE_MMIO_GET(dev) + BCM2711_MU_CNTL);
+
+	val = BCM2711_MU_CNTL_RX_ENABLE;
+	sys_write32(val, cntl);
+}
+
+static void uart_bcm2711_irq_rx_disable(const struct device *dev)
+{
+	uint32_t val = 0;
+	mem_addr_t cntl = (DEVICE_MMIO_GET(dev) + BCM2711_MU_CNTL);
+
+	val = (~BCM2711_MU_CNTL_RX_ENABLE);
+	sys_write32(val, cntl);
+}
+
+static int uart_bcm2711_irq_rx_ready(const struct device *dev)
+{
+	return bcm2711_mu_lowlevel_can_getc(DEVICE_MMIO_GET(dev));
```

Comment:
It is unnecessarily verbose having to do `DEVICE_MMIO_GET` in every function. Just do it once and same the address somewhere easily accessible (possibly the data structure).

Reply:
However, using `DEVICE_MMIO_GET `would improve readability.

Reply:
yeah, once. Having that all over the file it's just annoying.